nats-base-client/servers.ts

```diff
@@ -105,6 +117,45 @@ export class Servers {
     this.currentServer = this.servers[0];
   }
 
+  async resolve(): Promise<void> {
+    // If the client provides an urlParseFn, we are running in
+    // websockets. Since we don't handle any of the certificates
+    // changing to the IP without allowing the internal resolution
+    // will reject. So resolve is exclusive of an urlParseFn.
+    if (getUrlParseFn()) {
+      throw new Error("'resolve' option is not available to websocket clients");
+    }
+    const resolver = getResolveFn();
+    if (typeof resolver !== "function") {
+      throw new Error("'resolve' option is not available in the client");
+    }
+
+    const buf: ServerImpl[] = [];
+    for (const s of this.servers) {
+      if (isIP(s.hostname)) {
+        buf.push(s);
+        continue;
+      }
+      const ips = await resolver(s.hostname);
+      for (const ip of ips) {
+        // letting URL handle the details of representing IPV6 ip with a port, etc
+        const url = new URL(`http://${s.listen}`);
+        if (!url.port) {
+          url.port = `${DEFAULT_PORT}`;
+        }
+        url.hostname = ip;
+        const ss = new ServerImpl(url.host, false);
+        ss.tlsName = s.hostname;
+        buf.push(ss);
+      }
+    }
+    this.servers.length = 0;
+    if (this.randomize) {
+      shuffle(buf);
```

Comment:
Is `connect()` invoked only once? If not, are you not possibly re-shuffle the list of URLs that was already shuffled?

Also, remember that the client parity request for host name resolution was that: for a given URL in the list of servers, the library should resolve the hostname and, if getting multiple IPs, shuffle that list, then try to connect with IPs from that list until one succeeds (to TCP connect, not NATS connect).

In other words, suppose the user gives this list of servers as: A, B, C. Say library shuffled that to: C, A, B.
When connecting, you do: use "C", it is a host name, resolve, say you get IP_C only. Use that one. If fail to connect, move to A, say resolution returns IP_A1, IP_A2, IP_A3, shuffle that list, then use the first, if it TCP connects OK, then continue connection process, otherwise, go to the next in the "IP_A" list, etc..
But for the "servers" list, the "C, A, B" is shuffled once, then it is a matter of moving the tried server to the back. After trying "C", the list becomes "A, B, C", then after trying "A", "B, C, A", etc..

Reply:
The updated implementation does exactly this. It is actually simpler.